File: frontend/packages/gateway/api/src/refusal.ts
import surfacedFindings from './surfaced-findings.json';

/**
 * Which kind of refusal an answer is.
 *
 * The first eleven are the server's own, and the whole set is named here for the
 * reason the server names it: a caller writes a branch per kind, and a kind it
 * has never heard of is one it falls off the end of. Adding one on the server
 * is adding a case here.
 *
 * The last two are this client's, minted where no answer of the server's shape
 * arrived at all. They are kept in the same union so that everything a caller
 * has to handle is one type: a page that has to say why it is empty does not
 * care whether the sentence came from the server or from here.
 */
export type RefusalKind =
  | 'bad_path'
  /** The request itself was not one the route could read. */
  | 'bad_request'
  /**
   * The server does not answer this caller: the request showed no key of its,
   * arrived by a name that is not the server's own, or came from a page on
   * another site. It is refused before any route sees it, so it says nothing
   * about the Library.
   */
  | 'unauthorized'
  | 'no_such_entry'
  /**
   * The server answers nothing at the path that was asked, or answers that path
   * by other methods than the one it was asked by — `404` and `405`, one kind
   * between them. It is this server replying, which is why it is not
   * `unrecognized`: a page of an older build asking for a route since renamed
   * meets this, and the sentence names neither the path nor the method.
   */
  | 'no_such_route'
  | 'declined'
  /**
   * This device has to be enrolled in the Library again: a Master Key epoch was
   * activated, and the device holds only the key it replaced. Nothing a retry
   * mends, and nothing a page can do — enrolling happens at a terminal with the
   * new Recovery Code — so the sentence is all a screen has to show: a retry
   * offered beside it can only meet it again.
   */
  | 'epoch'
  /**
   * The server is locked, so nothing that needs the Master Key can be done: the
   * Passphrase is required, and the message says how to give it.
   *
   * The opposite verdict to `unauthorized`, and about the opposite person. That
   * one is said to somebody who is not the owner of this Library and tells them
   * nothing; this is said to the owner about their own device and tells them
   * everything. Nothing on a page can undo it — the Passphrase is typed at a
   * terminal — so what a screen does with it is show the sentence.
   *
   * Not to be read as the `locked` in {@link DeclinedReason}, which is one
   * Entry whose Container the Library records no key for and which no
   * Passphrase resolves. The two never arrive together: a locked server
   * declines nothing, because it fetches nothing.
   */
  | 'locked'
  | 'storage'
  | 'unverified'
  | 'server'
  /**
   * The request got no answer here: nothing is listening, the network went, or
   * the transfer broke while the body was still going up — which is
   * not proof the server never answered. Every `fetch` that rejects becomes
   * this, whatever it rejected for, except one the caller aborted: that is not
   * a refusal at all and passes through as itself.
   */
  | 'unreachable'
  /** Something answered, and it was not one of the shapes above. */
  | 'unrecognized';

/**
 * Which way something was declined, where it was.
 *
 * The first six are a fetch's, and a drop meets `unmapped`,
 * `unmaterializable`, `reserved` and `refused_root` as well. The last is a
 * drop's alone: the Library holds an Entry at that path inside a Pack, and
 * coffret cannot replace one of those yet — so the file is refused rather than
 * written where no sync could carry it in.
 */
export type DeclinedReason =
  | 'unmapped'
  | 'unmaterializable'
  /**
   * The path carries a name coffret keeps for itself inside a mapped folder:
   * the management area its own bookkeeping lives in, or the scratch a
   * half-written file is called by. A scan passes both over, so a file placed
   * under either would sit in the folder and never reach the Library.
   *
   * The management area's name is matched with ASCII case folded, and a
   * spelling that only folds to it is refused rather than passed over: a drop,
   * a read of one file, and a listing of a folder holding one are the three
   * declined this way. A sync or a freeze that stops at such a name is not one
   * of them — it reaches a page as a run that stopped, with no name in it. So
   * the name may be `.COFFRET` rather than `.coffret`, and it may stand in the
   * folder that was asked for rather than in the path itself.
   */
  | 'reserved'
  /**
   * A folder this device maps is not the folder its mapping was recorded
   * against — a copied disk, a mount that came back different — so nothing was
   * placed into it. Nothing on a page settles it: the message names the gesture,
   * and it is one at a terminal.
   */
  | 'refused_root'
  | 'surfaced'
  | 'locked'
  | 'pack_resident';

/**
 * The finding about one Entry, by the name the device layer gives it: what a
 * declined fetch reported, and what a run's finding reports in its field of the
 * same name.
 *
 * The last two only a sync finds, so they arrive on a finding and never on a
 * refusal: nothing a fetch does is declined over a file that changed inside a
 * Pack or one this device no longer has.
 */
export type SurfacedFinding =
  | 'ForeignFile'
  | 'LocallyChanged'
  | 'WitnessedDeletion'
  /**
   * A folder on the way to where the file belongs is not a folder of the mapped
   * folder — a symbolic link, or an ordinary file standing where a folder must
   * be. The rest of a run is unaffected: this is the shape of one folder.
   */
  | 'UnreachablePlace'
  | 'KeyLost'
  /**
   * The Entry's path carries `.coffret`, or a name differing from it only in
   * ASCII case, which is the name reserved for the device's own folder inside a
   * mapped folder at any depth. A placement refuses the two alike, and is the
   * one place they are one finding: a scan steps over the exact name and stops
   * at a fold of it. Nothing is placed there: a file
   * under it is one no later scan looks at, and one at the marker inside it
   * would take the mapped folder's identity away. No scan of this device makes
   * such a path, so it came from whichever device committed it.
   */
  | 'ReservedComponent'
  /** The file changed, and the Entry it changed from is inside a Pack. */
  | 'ChangedInPack'
  /** This device had the file and it is gone; the Library still holds it. */
  | 'DeletedLocally';

/**
 * Everything that can come back instead of an answer, in one shape.
 *
 * An `Error` so that it travels the way a failed request already does — thrown
 * out of the call, caught where the screen decides what to show — and a typed
 * one so that the screen can branch. `message` is the server's own sentence and
 * is written to be read by a person, which is why it is the one thing a caller
 * may display verbatim.
 */
export class Refusal extends Error {
  /** Which kind of refusal this is, for the caller to branch on. */
  readonly kind: RefusalKind;
  /** The HTTP status, and `0` where no answer arrived at all. */
  readonly status: number;
  /** Present exactly where the kind is `declined`. */
  readonly reason: DeclinedReason | null;
  /** Present where the reason is `surfaced` or `locked`. */
  readonly surfaced: SurfacedFinding | null;

  constructor(
    kind: RefusalKind,
    status: number,
    message: string,
    reason: DeclinedReason | null = null,
    surfaced: SurfacedFinding | null = null,
    options?: ErrorOptions,
  ) {
    super(message, options);
    this.name = 'Refusal';
    this.kind = kind;
    this.status = status;
    this.reason = reason;
    this.surfaced = surfaced;
  }
}

/** Whether something thrown out of this client is one of its refusals. */
export function isRefusal(thrown: unknown): thrown is Refusal {
  return thrown instanceof Refusal;
}

/**
 * The refusal one non-2xx answer stands for.
 *
 * It never throws, whatever came back. A body that is not the server's JSON is
 * an ordinary thing to receive — a proxy's own error page stands where the
 * server would have been — and a parser that threw there would replace a
 * refusal a caller can show with one it cannot.
 */
export async function refusalOf(response: Response): Promise<Refusal> {
  const body = await parsed(response);
  if (body === null) {
    // Said the way the person meets it: they asked the coffret server and no
    // coffret answer came back — whoever wrote this page, a proxy most of the
    // time, is not who they were asking. The status stays in the sentence
    // because it is the one clue for whoever then goes looking.
    return new Refusal(
      'unrecognized',
      response.status,
      `the coffret server did not answer — something else replied ${response.status} in its place`,
    );
  }
  return new Refusal(
    kindOf(body.error),
    response.status,
    body.message,
    reasonOf(body.reason),
    surfacedOf(body.surfaced),
  );
}

/** What a refusal looks like on the wire. */
interface RefusalBody {
  error: string;
  message: string;
  reason?: string;
  surfaced?: string;
}

/** The body as the server's refusal shape, or `null` where it is not one. */
async function parsed(response: Response): Promise<RefusalBody | null> {
  let body: unknown;
  try {
    body = await response.json();
  } catch {
    return null;
  }
  if (typeof body !== 'object' || body === null) {
    return null;
  }
  const fields = body as Record<string, unknown>;
  if (typeof fields.error !== 'string' || typeof fields.message !== 'string') {
    return null;
  }
  return {
    error: fields.error,
    message: fields.message,
    reason: typeof fields.reason === 'string' ? fields.reason : undefined,
    surfaced: typeof fields.surfaced === 'string' ? fields.surfaced : undefined,
  };
}

const KINDS: readonly string[] = [
  'bad_path',
  'bad_request',
  'unauthorized',
  'no_such_entry',
  'no_such_route',
  'declined',
  'epoch',
  'locked',
  'storage',
  'unverified',
  'server',
];

const REASONS: readonly string[] = [
  'unmapped',
  'unmaterializable',
  'reserved',
  'refused_root',
  'surfaced',
  'locked',
  'pack_resident',
];

/**
 * The finding names the server can send, read from the one file that holds
 * them.
 *
 * Not written out here, because a list written out here is a copy: the names
 * are the server's, spelled in its own `name_of` for a refusal and in
 * `finding.rs`'s `named` for a run's finding, and a variant renamed there
 * would leave this reading the new name as `null` and every screen showing the
 * generic sentence with nothing to say something had gone wrong. The file is
 * what the cases in `coffret-server` build from those `match`es and compare
 * against — a refusal's names at its head, a finding's the whole of it — so a
 * rename fails `cargo test` until the file is brought along, and this follows
 * the file with no second list to forget.
 *
 * {@link SurfacedFinding} stays written out, and is not a second copy of this:
 * it is the compile-time shape a caller branches on, checked where the branch
 * is written. What nothing checks is that union against the file — the file is
 * an array of strings, so a name it grew and the union did not is cast into the
 * union below and reaches a `switch` with no case for it. So a finding renamed
 * on the server is renamed in the file and in the union together, and each
 * case in `coffret-server` that compares its names with the file asks for both
 * where it fails.
 */
const FINDINGS: readonly string[] = surfacedFindings;

/**
 * The kind the body named, and `unrecognized` for one this client has not heard
 * of.
 *
 * A server that grew a kind is not a server this client can branch on, and
 * saying so is better than passing a string on as though it were one of the
 * eleven: a caller matching on the union would then fall through every case.
 */
function kindOf(named: string): RefusalKind {
  return KINDS.includes(named) ? (named as RefusalKind) : 'unrecognized';
}

function reasonOf(named: string | undefined): DeclinedReason | null {
  return named !== undefined && REASONS.includes(named) ? (named as DeclinedReason) : null;
}

function surfacedOf(named: string | undefined): SurfacedFinding | null {
  return named !== undefined && FINDINGS.includes(named) ? (named as SurfacedFinding) : null;
}

File: backend/crates/apps/coffret-server/src/routes/upload/refusal.rs
use axum::extract::multipart::MultipartError;
use axum::http::StatusCode;
use coffret_device::Error;

use crate::api_error::ApiError;

/// What a part was not taken for, and how far that reaches.
pub(super) enum Refusal {
    /// About this file alone, and the drop goes on: one file the Library will
    /// not have is not the file beside it.
    Part(ApiError),
    /// About the request, which stops here: whatever made it is no truer of the
    /// next part than of this one, so there is nothing to be gained by reading
    /// the rest. A budget this request has outrun, room this device has not for
    /// what is still coming or could not be asked about at all, a stream that
    /// broke before the next part, the state of a mapped root the parts are
    /// going through — one condition gathers them, and it is what the next of
    /// them is judged by.
    ///
    /// That condition is what the variant is chosen by, rather than the wire
    /// kind the refusal goes out under: a refused root reaches a page as
    /// `declined` the way a declined placement does (spec: EP-13), and a root
    /// this device could not read its own marker in reaches it as `server` the
    /// way this machine's other failures do — and both are about the whole
    /// drop.
    Request(ApiError),
    /// Not a refusal at all: the body stopped arriving while this part was
    /// being read.
    ///
    /// Nothing about the part or the request was decided, so neither of the
    /// variants above says it truly. What it is instead is the failure the
    /// route's own loop meets when the stream breaks between two parts, met a
    /// little earlier — so it is handed back to that loop as the failure it is,
    /// and the loop says it the one way it says its own. A refusal of this part
    /// would be recorded as a part somebody refused, which nobody did.
    Interrupted(MultipartError),
}

impl Refusal {
    /// What a failure to read the next chunk of a part comes to.
    ///
    /// Two different things arrive as the same error type, and axum's reading
    /// of the `multer` error under it is what tells them apart. A body that
    /// could not be read as multipart — a boundary that is not one, a part that
    /// ends before its data does — or one that ran past the body limit is the
    /// request being refused, a `400` or a `413` it has earned wherever in it
    /// that was found. A body whose stream failed — the connection went while a
    /// chunk was in flight — is `multer`'s `StreamReadFailed` over something
    /// that is not a limit, and axum reads that alone as `500`: nothing about
    /// what was sent is wrong with it, it stopped being sent.
    pub(super) fn reading(cause: MultipartError) -> Self {
        match cause.status() {
            StatusCode::INTERNAL_SERVER_ERROR => Self::Interrupted(cause),
            _ => Self::Request(ApiError::multipart(cause)),
        }
    }
}

/// Anything that goes wrong about one file is about that one file, so `?` on it
/// says so and the drop carries on.
///
/// Which is what keeps the other variant honest: an [`ApiError`]'s kinds are
/// the wire's, and they do not divide by reach — `declined` is what a refused
/// root and a refused part both go out under — so nothing in one tells this type
/// how far it reaches. This is therefore the conversion that can never make a
/// refusal about the request, and every place that decides the reach from the
/// request itself — a budget it has outrun, the room this device has, a body
/// that could not be read — writes [`Request`](Self::Request) out in full.
///
/// The conversion below is the one exception, and it is one because it has a
/// failure kind rather than a wire kind to read: a device error whose kind
/// settles the reach wherever it is met. So `?` on one of those may stop the
/// request, and its two arms are the whole of where that is decided.
impl From<ApiError> for Refusal {
    fn from(refusal: ApiError) -> Self {
        Self::Part(refusal)
    }
}

impl From<Error> for Refusal {
    fn from(cause: Error) -> Self {
        match cause {
            // The root the parts of this drop go into: in a drop onto a folder
            // it is one root for all of them, and at the Library root, where
            // the parts carry mappings of their own (spec: EP-9), it is the
            // first of those to be refused. Either way nothing is placed into
            // it and the request fails as a whole, the way a declined placement
            // fails a single writer's (spec: EP-11, EP-13).
            //
            // And the same root when nothing could be learned about it: a
            // permission the process has not on its marker is a fact about the
            // folder every part is going through, settled before the first of
            // them was read. Reported as one part's business the drop would
            // read the next part, meet the refusal again, and report it once
            // per file for a condition none of them caused. It is not a verdict
            // on the mapping and nothing here makes it one: where a refused root
            // tells the page which mapping to record again, this one tells it
            // only that the server could not answer, and the folder the disk
            // would not answer about stays out of the page and out of the record
            // alike (spec: EL-1, EP-11, EP-13).
            refused @ (Error::RootRefused(_) | Error::RootUnvouched { .. }) => {
                Self::Request(refused.into())
            }
            other => Self::Part(other.into()),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::io;

    use axum::body::{Body, Bytes};
    use axum::extract::multipart::MultipartError;
    use axum::extract::{FromRequest, Multipart};
    use axum::http::Request;
    use futures_util::stream;
    use tokio::sync::mpsc;

    use super::Refusal;

    /// The opening of one part, as a browser sends it, and none of its end.
    const PART_HEAD: &str = "--b\r\nContent-Disposition: form-data; name=\"file\"; \
                             filename=\"a.txt\"\r\nContent-Type: text/plain\r\n\r\nsome bytes";

    /// What reading one part's chunks meets, once its headers have arrived and
    /// the body then does `then`: fails, or — given nothing — simply ends.
    ///
    /// Fed through a channel rather than laid out in advance, because the
    /// failure has to arrive after the part has been handed over: a body that
    /// fails before its first part's headers are read is a failure of the
    /// route's own loop, which is not what these cases are about.
    async fn met_reading(then: Option<io::Error>) -> MultipartError {
        let (send, receive) = mpsc::unbounded_channel::<Result<Bytes, io::Error>>();
        send.send(Ok(Bytes::from_static(PART_HEAD.as_bytes())))
            .expect("the body is still being read");
        let body = Body::from_stream(stream::unfold(receive, |mut receive| async move {
            receive.recv().await.map(|item| (item, receive))
        }));
        let request = Request::builder()
            .method("POST")
            .uri("/api/upload")
            .header("content-type", "multipart/form-data; boundary=b")
            .body(body)
            .expect("a multipart request is well formed");
        let mut parts = Multipart::from_request(request, &())
            .await
            .expect("a multipart request is taken as one");
        let mut part = parts
            .next_field()
            .await
            .expect("the part's headers arrived whole")
            .expect("there is a part");
        match then {
            Some(failure) => send
                .send(Err(failure))
                .expect("the body is still being read"),
            None => drop(send),
        }
        loop {
            match part.chunk().await {
                Ok(Some(_)) => {}
                Ok(None) => panic!("the part was meant to end in a failure, and ended"),
                Err(cause) => return cause,
            }
        }
    }

    // A connection that went while a chunk was in flight is not a refusal of
    // anything: nothing about the part or the request was decided. It is handed
    // back as what it is, so that the part is never recorded as refused.
    #[tokio::test]
    async fn a_body_that_stopped_arriving_mid_part_is_no_refusal() {
        let went = io::Error::new(io::ErrorKind::ConnectionReset, "the connection went");

        assert!(matches!(
            Refusal::reading(met_reading(Some(went)).await),
            Refusal::Interrupted(_)
        ));
    }

    // A body that arrived whole and ends before its part does is a request that
    // cannot be read, and that is a refusal of the request — the one it always
    // was.
    #[tokio::test]
    async fn a_body_that_ends_mid_part_is_a_refused_request() {
        match Refusal::reading(met_reading(None).await) {
            Refusal::Request(refusal) => assert_eq!(refusal.kind(), "bad_request"),
            _ => panic!("a truncated body is the request being refused"),
        }
    }
}
